[AArch64][GlobalISel] Improve legalization of G_PTR_ADD

davemgreen opened the pull request:
The testing we have for vector ptradd was a bit lacking. In adding tests this patch found a couple of issues mostly with the way v3 vectors of ptrs were sometimes legalized via i64, and with non-i64 additions. It does not attempt to fix the issue with mergevalues from returning vector ptrs.

tschuett commented:
Thanks!

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -212,7 +213,7 @@ void LegalizerHelper::mergeMixedSubvectors(Register DstReg,
 
   Register Leftover = PartRegs[PartRegs.size() - 1];
-  if (MRI.getType(Leftover).isScalar())
+  if (MRI.getType(Leftover).isScalar() || MRI.getType(Leftover).isPointer())
```
!isVector?

tschuett commented on llvm/test/CodeGen/AArch64/ptradd.ll:
```diff
@@ -0,0 +1,221 @@
+; RUN: llc < %s -mtriple=aarch64-none-eabi -global-isel | FileCheck %s --check-prefixes=CHECK,CHECK-GI
+
+; Note: these tests use stores instead of returns as the return handling for
+; vector ptrs is currently sometimes create invalid merge values.
```
Nit: unmerge values.

davemgreen committed:
[AArch64][GlobalISel] Improve legalization of G_PTR_ADD

The testing we have for vector ptradd was a bit lacking. In adding tests this
patch found a couple of issues mostly with the way v3 vectors of ptrs were
sometimes legalized via i64, and with non-i64 additions.
